repair/row_level.cc

```diff
@@ -979,6 +998,42 @@ class repair_meta {
         return with_gate(_gate, [this, estimated_partitions] {
             _estimated_partitions = estimated_partitions;
             _repair_writer->set_estimated_partitions(_estimated_partitions);
+            _repair_reader.emplace(_db,
+                _cf,
+                _schema,
+                _permit,
+                _range,
+                _remote_sharder,
+                _master_node_shard_config.shard,
+                _seed,
+                std::invoke([this]() {
+                    if (_repair_master || _same_sharding_config) {
+                        return repair_reader::read_strategy::local;
+                    }
+
+                    // multishard_filter means load all the data in the range and apply
+                    // filter by master shard on top, discarding partitions from other shards.
+                    // multishard_split means split the range into multiple subranges,
+                    // each containing only the required data from the master shard.
+                    // For situations with a sparse data set spread across numerous ranges,
+                    // the overhead from continuously switching between these ranges,
+                    // specifically during the fast_forward_to function on the multishard_reader,
+                    // can become the main factor in performance.
+                    // Similarly, with multishard_filter, reading all the partitions within
+                    // the range can lead to the next_partition cost dominating the overall cost.
+                    // The heuristic here chooses the strategy with minimal such cost.
+                    // Note that with multishard_filter we don't read entire partitions which
+                    // can be a huge waste. We only fill the buffer inside
+                    // flat_mutation_reader_v2, if a partition is found to belong to the incorrect
+                    // master shard, we call next_partition(), which effectively clears
+                    // the buffer until the next partition is reached.
+                    if (_local_range_estimation &&
+                        _local_range_estimation->partitions_count <= _local_range_estimation->master_subranges_count)
```

Comment:
Let's add a debug log about the partitions_count,  master_subranges_count and the read strategy selection. 

Perhaps we can relax the check a bit, e.g., 


_local_range_estimation->partitions_count * factor <= _local_range_estimation->master_subranges_count

factor = 0.5 or 0.8

Reply:
I added the debug logs.

I'm not quite into the `factor` thing though, it seems to blur the heuristic reasoning outlined in the comment.

Reply:
How about adding the keyspace={}, table={}, range={}, as well. You can get it with: _schema->ks_name()  _schema->cf_name().

Reply:
The factor thing is like saying if _local_range_estimation->master_subranges_count / _local_range_estimation->partitions_count >= factor, we apply the filter reader trick. 

It does not have to be 1. The benefits of the filter reader probably shows up with a smaller factor already. 

With avi's unit test suggestions, we can play with different configurations to find a better empirical factor.

Reply:
> How about adding the keyspace={}, table={}, range={}, as well. You can get it with: _schema->ks_name() _schema->cf_name().

ok. I noticed that most of the `debug/trace` logs in `row_level.cc` lacks this 'context-binding' information, such as these table/range or just `repair_meta_id`. It might make it difficult to reason about and draw correlations. Perhaps we should revisit that in separate PR.

Reply:
> The factor thing is like saying if _local_range_estimation->master_subranges_count / _local_range_estimation->partitions_count >= factor, we apply the filter reader trick.
> 
> It does not have to be 1. The benefits of the filter reader probably shows up with a smaller factor already.
> 
> With avi's unit test suggestions, we can play with different configurations to find a better empirical factor.

I looked at the actual values which `get_estimated_partitions` returns, in my test there were 300 partitions spread around two shards, master range is `open_ended_both_sides`, subranges=4000, estimated partitions=126298. This is because of the accumulated error, 
total_keys=300 total_pages=16 estimated_keys_for_subrange=16, total =16 *4000=64000, which is multiplied by two shards.

The point it, `get_estimated_partitions` doesn't work if the number of subranges is big, so currently there is no point in tuning the correlation with additional coefficients.

I don't have better idea for this heuristic. I left it as is, since it still solves the initial problem with slow repair in debug mode - `get_estimated_partitions` returns zero for most of the ranges and we use fast `multishard_filter`.